logstash-core/lib/logstash/logging/slow_logger.rb

```diff
@@ -0,0 +1,42 @@
+# encoding: utf-8
+require "logstash/namespace"
+require "logstash/logging/slow_logger/freq_items"
+
+module LogStash
+  module Logging
+
+    class NullLogger
+      def log(threshold, time, data={})
+      end
+      alias_method :warn, :log
+    end
+
+    class SlowLogger
+
+      attr_reader :logger, :freq_items
+
+      def initialize(name="loggers.slow", params={})
+        @logger     = LogStash::Logging::Logger.new(name)
+        @freq_items = LogStash::Logging::Util::FreqItems.new
+      end
+
+      def log(threshold, time, data={})
+        message          = data.delete(:message)
+        data[:threshold] = threshold
+        data[:time]      = time
+
+        cache_elements(threshold, time)
+        logger.warn(message, data)
```

Comment:
I think this `logger.warn(message, data)` is not a great logging format. Can we have more structured logging here? Have explicit: 
`took_millis`
`plugin` (which plugin?)
`method`
`event_payload` (if enabled)
`optional_message`

When you use the `json` format to log, these should be nice fields for the user to use in ES (in the future)